File: linked_lists/sum_lists/sol1_f.py
import sys

sys.path.append('../')

from sll import sll

def paddling(slinked_list, num):
  key = slinked_list.size
  for i in xrange(num):
    slinked_list.addFirst(key, 0)

def psum(nd1, nd2):
  if not nd1 and not nd2:
    return sll._node(0, 0)
    
  nd_n = psum(nd1.next, nd2.next)
  
  data = nd1.data + nd2.data + nd_n.data
  carry = data // 10
  data = data % 10
  
  nd_n.data = data
  
  nd_c = sll._node(carry, 0)
  nd_c.next = nd_n
  
  return nd_c


def sum_lists(sll1, sll2):
  if not sll1:
    return sll2
    
  if not sll2:
    return sll1
    
  if sll1.size > sll2.size:
    paddling(sll2, sll1.size - sll2.size)
  else:
    paddling(sll1, sll2.size - sll1.size)
    
  nd1 = sll1.head
  nd2 = sll2.head
    
  nd = psum(nd1, nd2)
  
  sll3 = sll.sll()
  sll3.head = nd
    
  return sll3
  
if __name__ == "__main__":
  slinked_list1 = sll.sll()
  key = 0
  value = [7, 1, 6]
  for it in value:
    slinked_list1.addLast(key, it)
    key += 1
  slinked_list2 = sll.sll()
  key = 0
  value = [5, 9, 2]
  for it in value:
    slinked_list2.addLast(key, it)
    key += 1
    
  print slinked_list1
  print slinked_list2
  
  print "add"
  
  print sum_lists(slinked_list1, slinked_list2)

File: linked_lists/sum_lists/sol1_f.cpp
#include "sll.h"

#include <iostream>
#include <vector>

unsigned int linkedListSize(node* nd) {
  unsigned int size = 0;
  while ( nd ) {
     ++size;
     nd = nd->next();
  }
  return size;
}

node* addNode(node* nd1, node* nd2, int& add) {
  int res = nd1->data() +add;
  if ( nd2 != nullptr ) {
     res +=  nd2->data();
  }
  node* resND = new node(res % 10, res);
  add = res / 10;
  return resND;
}

node* sumListsNode(node* nd1, node* nd2, unsigned int k, int& add) {
  node* resND, *addcurrent;
  
  if ( nd1->next() == nullptr && nd2->next() == nullptr ) {
    return addNode(nd1, nd2, add);
  }
  
  if ( k > 0 ) {
    resND = sumListsNode(nd1->next(), nd2, --k, add);
    addcurrent = addNode(nd1, nullptr, add);
    addcurrent->setNext(resND);
  } else {
    resND = sumListsNode(nd1->next(), nd2->next(), k, add);
    addcurrent = addNode(nd1, nd2, add);
    addcurrent->setNext(resND);
  }
  return addcurrent;
}

node* sumLists(node* nd1, node* nd2) {
   unsigned int size1 = linkedListSize(nd1);
   unsigned int size2 = linkedListSize(nd2);
   if ( size1 < size2 ) {
      return sumLists(nd2, nd1);
   }
   
   int add = 0;
   node* head = sumListsNode(nd1, nd2, size1 - size2, add);
   if ( add == 1 ) {
     node* tmp = new node(add, 0);
     tmp->setNext(head);
     head = tmp;
   }
   
   return head;
}


int main() {
  sll slinked_list;
  sll slinked_list1;
  std::vector<int> value{1, 2, 3, 2, 3};
  std::vector<int> value1{9, 2, 1};
  int key = 0;
  for (auto& it : value) {
    std::cout << key << std::endl;
    if (key == 0) {
      slinked_list.insertHead(it, key);
    } else {
      slinked_list.insertLast(it, key);
    }
    ++key;
  }
  
  for (auto& it : value1) {
    std::cout << key << std::endl;
    if (key == 0) {
      slinked_list1.insertHead(it, key);
    } else {
      slinked_list1.insertLast(it, key);
    }
    ++key;
  }
  
  node* nd = slinked_list.head();
  while (nd) {
    std::cout << nd->data() << " ";
    nd = nd->next();
  }
  std::cout << std::endl;
   
  nd = slinked_list1.head();
  while (nd) {
    std::cout << nd->data() << " ";
    nd = nd->next();
  }
  std::cout << std::endl;
   
  nd = slinked_list.head();
  node* nd1 = slinked_list1.head();
  nd = sumLists(nd, nd1);
  
  std::cout << "getting result: ";
  while (nd) {
    std::cout << nd->data() << " ";
    nd = nd->next();
  }
  std::cout << std::endl;
}